### PR TITLE
Add PodTemplateSpec in cassdc crd spec

### PR DESCRIPTION
**Pull request by sandoichi**

### CODE REVIEW

File: operator/pkg/reconciliation/constructor.go [hunk 1]
```diff
@@ -461,3 +463,46 @@ func calculatePodAntiAffinity(allowMultipleNodesPerWorker bool) *corev1.PodAntiA
...
+	merged := baseTemplate.DeepCopy()
+
+	merged.Labels = utils.MergeMap(userTemplate.Labels, baseTemplate.Labels)
+	merged.Spec.Volumes = append(merged.Spec.Volumes, userTemplate.Spec.Volumes...)
+	merged.Spec.InitContainers = append(merged.Spec.InitContainers, userTemplate.Spec.InitContainers...)
+	merged.Spec.Containers = append(merged.Spec.Containers, userTemplate.Spec.Containers...)
+	merged.Spec.RestartPolicy = userTemplate.Spec.RestartPolicy
+	merged.Spec.TerminationGracePeriodSeconds = userTemplate.Spec.TerminationGracePeriodSeconds
+
+	merged.Spec.ActiveDeadlineSeconds = userTemplate.Spec.ActiveDeadlineSeconds
```
[the remark this answers is not in the document]
Comment (jimdickinson) [reply operator/pkg/reconciliation/constructor.go:483]: I agree with John, with the exception that slices we setup should be prepended to user provided slices.

File: operator/pkg/reconciliation/constructor.go [hunk 2]
```diff
@@ -461,3 +291,169 @@ func calculatePodAntiAffinity(allowMultipleNodesPerWorker bool) *corev1.PodAntiA
...
+		serviceAccount = dc.Spec.ServiceAccount
+	}
+	baseTemplate.Spec.ServiceAccountName = serviceAccount
+
+	// init containers
+	initContainers, err := buildInitContainers(dc)
+	if err != nil {
+		return nil, err
+	}
+	baseTemplate.Spec.InitContainers = append(baseTemplate.Spec.InitContainers, initContainers...)
```
Comment (johntrimble) [inline operator/pkg/reconciliation/constructor.go:443]: I think Jim wanted the init containers prepended to the ones the user provides, same with the containers.

Comment (sandoichi) [reply operator/pkg/reconciliation/constructor.go]: whoops... should be fixed now

File: operator/pkg/reconciliation/constructor.go [hunk 3]
```diff
@@ -461,3 +291,172 @@ func calculatePodAntiAffinity(allowMultipleNodesPerWorker bool) *corev1.PodAntiA
 		},
 	}
 }
+
+func selectorFromFieldPath(fieldPath string) *corev1.EnvVarSource {
```
Comment (jimdickinson) [inline operator/pkg/reconciliation/constructor.go:295]: 💯 I'd thought about doing this before

**Comment by bradfordcp** [comment]
Does this change remove nodeSelector from CassandraDatacenterSpec?

Would we deprecate that field and favor the one present in the PodTemplateSpec?

File: operator/pkg/reconciliation/constructor.go [hunk 4]
```diff
@@ -461,3 +291,172 @@ func calculatePodAntiAffinity(allowMultipleNodesPerWorker bool) *corev1.PodAntiA
...
+		{Name: "PRODUCT_NAME", Value: dc.Spec.ServerType},
+		// TODO remove this post 1.0
+		{Name: "DSE_VERSION", Value: serverVersion},
+	}
+
+	return []corev1.Container{serverCfg}, nil
+}
+
+func buildPodTemplateSpec(dc *api.CassandraDatacenter, zone string, rackName string) (*corev1.PodTemplateSpec, error) {
+	baseTemplate := dc.Spec.PodTemplate
```
Comment (jimdickinson) [inline operator/pkg/reconciliation/constructor.go:398]: do we need to `DeepCopy()` here?

File: .gitignore [hunk 5]
```diff
@@ -94,6 +94,7 @@ gradle-app.setting
 .gogradle
 
 operator/vendor
+vendor
```
Comment (sandoichi) [inline .gitignore:97]: now that we have a single go.mod in the root...i've been getting `vendor` dirs created in the root on occasion

Comment (jimdickinson) [reply .gitignore:97]: merge master in and you'll get #88

File: operator/pkg/reconciliation/constructor.go [hunk 6]
```diff
@@ -461,3 +291,172 @@ func calculatePodAntiAffinity(allowMultipleNodesPerWorker bool) *corev1.PodAntiA
...
+	}
+	serverVersion := dc.Spec.ServerVersion
+	serverCfg.Env = []corev1.EnvVar{
+		{Name: "CONFIG_FILE_DATA", Value: configData},
+		{Name: "POD_IP", ValueFrom: selectorFromFieldPath("status.podIP")},
+		{Name: "RACK_NAME", ValueFrom: selectorFromFieldPath(fmt.Sprintf("metadata.labels['%s']", api.RackLabel))},
+		{Name: "PRODUCT_VERSION", Value: serverVersion},
+		{Name: "PRODUCT_NAME", Value: dc.Spec.ServerType},
+		// TODO remove this post 1.0
+		{Name: "DSE_VERSION", Value: serverVersion},
```
Comment (johntrimble) [inline operator/pkg/reconciliation/constructor.go:391]: Do we still need this? The TODO says "remove this post 1.0" and we are post 1.0. Also, config-builder doesn't even reference it now: 

https://github.com/datastax/cass-config-builder/blob/master/src/com/datastax/configbuilder/operator.clj#L60-L68

File: tests/podspec_simple/podspec_simple_suite_test.go [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+				WithFlag("field-selector", "status.phase=Running").
+				FormatOutput(json)
+			ns.WaitForOutputAndLog(step, k, "true", 120)
+
+			step = "creating a datacenter resource with 1 racks/1 nodes"
+			k = kubectl.ApplyFiles(dcYaml)
+			ns.ExecAndLog(step, k)
+
+			step = "waiting for the node to become ready"
+			json = `jsonpath={.items[*].status.containerStatuses[?(@.name=="cassandra")].ready}`
```
Comment (johntrimble) [inline tests/podspec_simple/podspec_simple_suite_test.go:63]: :+1:

File: operator/pkg/reconciliation/constructor.go [hunk 8]
```diff
@@ -461,3 +291,170 @@ func calculatePodAntiAffinity(allowMultipleNodesPerWorker bool) *corev1.PodAntiA
...
+		{Name: "POD_IP", ValueFrom: selectorFromFieldPath("status.podIP")},
+		{Name: "RACK_NAME", ValueFrom: selectorFromFieldPath(fmt.Sprintf("metadata.labels['%s']", api.RackLabel))},
+		{Name: "PRODUCT_VERSION", Value: serverVersion},
+		{Name: "PRODUCT_NAME", Value: dc.Spec.ServerType},
+	}
+
+	return []corev1.Container{serverCfg}, nil
+}
+
+func buildPodTemplateSpec(dc *api.CassandraDatacenter, zone string, rackName string) (*corev1.PodTemplateSpec, error) {
```
Comment (jimdickinson) [inline operator/pkg/reconciliation/constructor.go:395]: let's add a unit test here? this signature is great

File: operator/pkg/reconciliation/constructor.go [hunk 9]
```diff
@@ -461,3 +291,170 @@ func calculatePodAntiAffinity(allowMultipleNodesPerWorker bool) *corev1.PodAntiA
...
+
+	configData, err := dc.GetConfigAsJSON()
+	if err != nil {
+		return nil, err
+	}
+	serverVersion := dc.Spec.ServerVersion
+	serverCfg.Env = []corev1.EnvVar{
+		{Name: "CONFIG_FILE_DATA", Value: configData},
+		{Name: "POD_IP", ValueFrom: selectorFromFieldPath("status.podIP")},
+		{Name: "RACK_NAME", ValueFrom: selectorFromFieldPath(fmt.Sprintf("metadata.labels['%s']", api.RackLabel))},
```
Comment (jimdickinson) [inline operator/pkg/reconciliation/constructor.go:387]: do you think we should just pass the rackName as a simple string? unrelated to your PR, but I've thought this before and now we're touching the code

File: operator/pkg/reconciliation/constructor.go [hunk 10]
```diff
@@ -461,3 +291,170 @@ func calculatePodAntiAffinity(allowMultipleNodesPerWorker bool) *corev1.PodAntiA
...
+
+	serverImage, err := dc.GetServerImage()
+	if err != nil {
+		return nil, err
+	}
+
+	cassContainer.Image = serverImage
+	cassContainer.Resources = dc.Spec.Resources
+	cassContainer.Env = []corev1.EnvVar{
+		{Name: "DS_LICENSE", Value: "accept"},
```
Comment (jimdickinson) [inline operator/pkg/reconciliation/constructor.go:329]: all these EnvVar structs are 10x more readable 👍

**Pull request merged**